src/dsl/dsl.jl

```diff
@@ -63,46 +64,21 @@ end
 include("dynamic.jl")
 include("static.jl")
 
-function address_from_expression(lhs)
-    if lhs isa Symbol
-        QuoteNode(lhs)
-    else
-        error("Syntax error: Only a variable or an address expression can appear on the lefthand side of a ~. Invalid left-hand side: $(lhs).")
-    end
-end
-
 function desugar_tildes(expr)
-    trace_ref = GlobalRef(@__MODULE__, Symbol("@trace"))
-    line_num = LineNumberNode(1, :none)
     MacroTools.postwalk(expr) do e
-        # Replace with globally referenced macrocalls
-        if MacroTools.@capture(e, {*} ~ rhs_)
-            Expr(:macrocall, trace_ref, line_num, rhs)
-        elseif MacroTools.@capture(e, {addr_} ~ rhs_)
-            Expr(:macrocall, trace_ref, line_num, rhs, addr)
+        # Replace tilde statements with :gentrace expressions
+        if MacroTools.@capture(e, {*} ~ rhs_call)
```

Comment:
Nope, prior to macro-expansion, `*` simply matches to an unqualified occurrence of `*`:
```julia
julia> dump(:({*}))
Expr
  head: Symbol braces
  args: Array{Any}((1,))
    1: Symbol *
```
If we want to match qualified occurrences of `*`, e.g., `A.B.Gen.*`, we have to add another branch in the if-then-else statement which does that specifically, and make sure that `A.B.Gen` is an appropriately qualified reference to the actual `Gen` module. I think we should avoid doing this and simply treat `{*}` as special syntax -- there's no use case in which we would want a `*` in the middle of two braces to refer to `Base.:*`.

Reply:
Hm.  Weird that `~` and `*` seem to be treated differently.

Reply:
They aren't! They're treated the same. How do you understand them as being treated differently?

Reply:
Ohh, after re-reading https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659008917 I now realize that the bug was in the macro that produced something that wasn't actually a plain `~`, i.e. that issue is independent of Gen.  Got it!

Reply:
Does that mean https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659657114 is wrong?  I.e. at what point during lexing, parsing, macro expansion, evaluation would `~` get turned into `Main.:~` if we didn't intervene?

Reply:
Macro-expansion itself turns `~` into `Main.:~`:

```julia
julia> macro draw_x()
         :( $(esc(:x)) ~ normal(0, 1) )
       end
@draw_x (macro with 1 method)
julia> dump(macroexpand(Main, :(@draw_x())))
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: GlobalRef
      mod: Module Main
      name: Symbol ~
    2: Symbol x
    3: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: GlobalRef
          mod: Module Main
          name: Symbol normal
        2: Int64 0
        3: Int64 1
```

You can see in the dumped Expression that `~` has become a `GlobalRef` with module `Main` and symbol `~`.

Reply:
It seems in that example that the resolution of `~` to `Main.:~` happened at macro definition time, not expansion time, right?  So macro expansion just substituted the already-resolved name.  I'd wonder what happens in an example like
```julia
macro draw_x()
  Expr(:~, esc(:x), normal(0, 1) )
end
```
but I don't know how to make an actually correct version of that example.

BTW, this is off-topic, so I understand if you don't have time to keep explaining this stuff :slightly_smiling_face:

Reply:
Oh, this is exactly what @georgematheos was saying in https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659008917 :slowbro:

Reply:
I'm definitely in favor of dropping the preemptive intercepting of `Module.:~` (unless there are further reasons we need it beyond the above described).  We'd be breaking legitimate (albeit weird with poor name choices) uses of Julia semantics.  It could be helpful to check for possible `~`-related mistakes and issue a `@warning` if found.  But I think hypercorrecting potentially-correct-but-weird programs is worse than crashing.

Reply:
This is exactly what @ztangent said in https://github.com/probcomp/Gen.jl/pull/280#issuecomment-659652523 :slowwwbrooo: